contracts/lib/Multicall.cairo

```diff
@@ -3,49 +3,94 @@
 from starkware.starknet.common.syscalls import call_contract, get_block_number
 from starkware.cairo.common.alloc import alloc
 from starkware.cairo.common.memcpy import memcpy
+from contracts.utils.calls import (
+    Call,
+    CallArray,
+    from_call_array_to_call,
+)
+
 
 // ////////////////////////////////////////////////////////////////////////
 // The Multicall contract can call an array of view methods on different
 // contracts and return the aggregate response as an array.
-// E.g.
-// Input: [to_1, selector_1, data_1_len, data_1, ..., to_N, selector_N, data_N_len, data_N]
-// Output: [result_1 + .... + result_N]
-// ////////////////////////////////////////////////////////////////////////
-
+// Input: same as the IAccount.__execute__ 
+// @return (block_number, retdata_size, retdata)
+//   Where retdata is [len(call_1_data), *call_1_data, len(call_1_data), *call_2_data, ..., len(call_N_data), *call_N_data]
+// ///////////////////////////////////////////////////////////////////////
 @view
-func aggregate{syscall_ptr: felt*, range_check_ptr}(calls_len: felt, calls: felt*) -> (
-    block_number: felt, result_len: felt, result: felt*
+func aggregate{syscall_ptr: felt*, range_check_ptr}(
+    call_array_len: felt,
+    call_array: CallArray*,
+    calldata_len: felt,
+    calldata: felt*
+) -> (
+    block_number: felt, retdata_len: felt, retdata: felt*
 ) {
     alloc_locals;
-
-    let (result: felt*) = alloc();
-    let (result_len) = call_loop(calls_len=calls_len, calls=calls, result=result);
+    let (retdata_len, retdata) = execute_call_array_detailed(call_array_len, call_array, calldata_len, calldata);
     let (block_number) = get_block_number();
+    return (block_number=block_number, retdata_len=retdata_len, retdata=retdata);
+}
+
+
+// @notice Convenience method to convert an execute a call array
+// @return response_len: The size of the returned data
+// @return response: Data return 
+//   in the form [len(call_1_data), *call_1_data, len(call_1_data), *call_2_data, ..., len(call_N_data), *call_N_data]
+func execute_call_array_detailed{syscall_ptr: felt*}(
```

Comment:
I don't like the use of the word `execute` in this context and I find it misleading. Because of the `__execute__` method of the account the word `execute` is associated to invoke calls that changes the state of the blockchain. Here we are aggregating `read` only calls so we should use something else.

I do prefer the use of `call_loop` as it was before which both shows it is a (message) call and also that it is a recursive method. Or we can use `do` instead of `execute` if you don't like it?

So maybe replace `execute_call_array_detailed` by either `call_arrays_loop` or `do_call_array_detailed`, and `execute_calls_detailed` by `call_loop` or `do_calls_detailed` ?

Reply:
renamed to `do_call_array` and `do_calls`